File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/models/ModelInfo.kt
package com.runanywhere.sdk.models

import com.runanywhere.sdk.data.models.fileExists
import com.runanywhere.sdk.models.enums.LLMFramework
import com.runanywhere.sdk.models.enums.ModelCategory
import com.runanywhere.sdk.models.enums.ModelFormat
import com.runanywhere.sdk.utils.SimpleInstant
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

/**
 * Information about a model - exact match with iOS ModelInfo
 */
@Serializable
data class ModelInfo(
    // Essential identifiers
    val id: String,
    val name: String,
    val category: ModelCategory,

    // Format and location
    val format: ModelFormat,
    val downloadURL: String? = null,
    var localPath: String? = null,

    // Size information (in bytes)
    val downloadSize: Long? = null,
    val memoryRequired: Long? = null,

    // Integrity verification
    val sha256Checksum: String? = null,
    val md5Checksum: String? = null,

    // Framework compatibility
    val compatibleFrameworks: List<LLMFramework> = emptyList(),
    val preferredFramework: LLMFramework? = null,

    // Model-specific capabilities (optional based on category)
    val contextLength: Int? = null,
    val supportsThinking: Boolean = false,

    // Optional metadata
    val metadata: ModelInfoMetadata? = null,

    // Tracking fields
    val source: ConfigurationSource = ConfigurationSource.REMOTE,
    val createdAt: SimpleInstant = SimpleInstant.now(),
    var updatedAt: SimpleInstant = SimpleInstant.now(),
    var syncPending: Boolean = false,

    // Usage tracking
    var lastUsed: SimpleInstant? = null,
    var usageCount: Int = 0,

    // Non-persistent runtime properties
    @Transient
    var additionalProperties: Map<String, String> = emptyMap()
) {
    /**
     * Whether this model is downloaded and available locally
     * Matches iOS: checks both localPath existence AND file existence on disk
     */
    val isDownloaded: Boolean
        get() {
            val path = localPath ?: return false

            // Built-in models are always available (like iOS)
            if (path.startsWith("builtin://") || path.startsWith("builtin:")) {
                return true
            }

            // Check if file/directory actually exists on disk (matches iOS behavior)
            return fileExists(path)
        }

    /**
     * Whether this model is available for use (downloaded and locally accessible)
     */
    val isAvailable: Boolean
        get() = isDownloaded

    /**
     * Get the effective context length (with defaults based on category)
     */
    val effectiveContextLength: Int?
        get() = if (category.requiresContextLength) {
            contextLength ?: 2048
        } else {
            contextLength
        }

    /**
     * Get the effective thinking support (based on category)
     */
    val effectiveSupportsThinking: Boolean
        get() = if (category.supportsThinking) supportsThinking else false
}

/**
 * Model information metadata - exact match with iOS
 */
@Serializable
data class ModelInfoMetadata(
    val author: String? = null,
    val license: String? = null,
    val tags: List<String> = emptyList(),
    val description: String? = null,
    val trainingDataset: String? = null,
    val baseModel: String? = null,
    val quantizationLevel: QuantizationLevel? = null,
    val version: String? = null,
    val minOSVersion: String? = null,
    val minMemory: Long? = null
)

/**
 * Quantization level for models
 */
@Serializable
enum class QuantizationLevel(val value: String) {
    Q2_K("q2_k"),
    Q3_K_S("q3_k_s"),
    Q3_K_M("q3_k_m"),
    Q3_K_L("q3_k_l"),
    Q4_0("q4_0"),
    Q4_1("q4_1"),
    Q4_K_S("q4_k_s"),
    Q4_K_M("q4_k_m"),
    Q5_0("q5_0"),
    Q5_1("q5_1"),
    Q5_K_S("q5_k_s"),
    Q5_K_M("q5_k_m"),
    Q6_K("q6_k"),
    Q6_K_L("q6_k_l"),
    Q8_0("q8_0"),
    F16("f16"),
    F32("f32");

    companion object {
        fun fromValue(value: String): QuantizationLevel? {
            return entries.find { it.value.equals(value, ignoreCase = true) }
        }
    }
}

/**
 * Configuration source
 */
@Serializable
enum class ConfigurationSource {
    LOCAL,
    REMOTE,
    DEFAULT
}

File: sdk/runanywhere-flutter/lib/core/models/model/model_info.dart
import 'dart:io';

import '../common/configuration_source.dart';
import '../common/thinking_tag_pattern.dart';
import '../framework/llm_framework.dart';
import '../framework/model_format.dart';
import 'model_category.dart';
import 'model_info_metadata.dart';

/// Information about a model
/// Matches iOS ModelInfo from Core/Models/Model/ModelInfo.swift
class ModelInfo {
  // Essential identifiers
  final String id;
  final String name;
  final ModelCategory category;

  // Format and location
  final ModelFormat format;
  final Uri? downloadURL;
  Uri? localPath;

  // Size information (in bytes)
  final int? downloadSize;
  final int? memoryRequired;

  // Framework compatibility
  final List<LLMFramework> compatibleFrameworks;
  final LLMFramework? preferredFramework;

  // Model-specific capabilities
  final int? contextLength;
  final bool supportsThinking;
  final ThinkingTagPattern? thinkingPattern;

  // Optional metadata
  final ModelInfoMetadata? metadata;

  // Tracking fields
  final ConfigurationSource source;
  final DateTime createdAt;
  DateTime updatedAt;
  bool syncPending;

  // Usage tracking
  DateTime? lastUsed;
  int usageCount;

  // Non-Codable runtime properties
  Map<String, String> additionalProperties;

  ModelInfo({
    required this.id,
    required this.name,
    required this.category,
    required this.format,
    this.downloadURL,
    this.localPath,
    this.downloadSize,
    this.memoryRequired,
    List<LLMFramework>? compatibleFrameworks,
    LLMFramework? preferredFramework,
    int? contextLength,
    bool supportsThinking = false,
    ThinkingTagPattern? thinkingPattern,
    this.metadata,
    ConfigurationSource? source,
    DateTime? createdAt,
    DateTime? updatedAt,
    this.syncPending = false,
    this.lastUsed,
    this.usageCount = 0,
    Map<String, String>? additionalProperties,
  })  : compatibleFrameworks = compatibleFrameworks ?? [],
        preferredFramework =
            preferredFramework ?? compatibleFrameworks?.firstOrNull,
        contextLength = category.requiresContextLength
            ? (contextLength ?? 2048)
            : contextLength,
        supportsThinking = category.supportsThinking ? supportsThinking : false,
        thinkingPattern = (category.supportsThinking && supportsThinking)
            ? (thinkingPattern ?? ThinkingTagPattern.defaultPattern)
            : null,
        source = source ?? ConfigurationSource.remote,
        createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now(),
        additionalProperties = additionalProperties ?? {};

  /// Whether this model is downloaded and available locally
  bool get isDownloaded {
    final path = localPath;
    if (path == null) return false;

    // Built-in models (e.g., Apple Foundation Models) are always available
    if (path.scheme == 'builtin') {
      return true;
    }

    // Check if the file or directory actually exists on disk
    final localFile = File(path.toFilePath());
    final localDir = Directory(path.toFilePath());

    if (localFile.existsSync()) {
      return true;
    }

    if (localDir.existsSync()) {
      // For directories, verify they contain files (not empty)
      final contents = localDir.listSync();
      return contents.isNotEmpty;
    }

    return false;
  }

  /// Whether this model is available for use (downloaded and locally accessible)
  bool get isAvailable => isDownloaded;

  /// JSON serialization
  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'category': category.rawValue,
        'format': format.rawValue,
        if (downloadURL != null) 'downloadURL': downloadURL.toString(),
        if (localPath != null) 'localPath': localPath.toString(),
        if (downloadSize != null) 'downloadSize': downloadSize,
        if (memoryRequired != null) 'memoryRequired': memoryRequired,
        'compatibleFrameworks':
            compatibleFrameworks.map((f) => f.rawValue).toList(),
        if (preferredFramework != null)
          'preferredFramework': preferredFramework!.rawValue,
        if (contextLength != null) 'contextLength': contextLength,
        'supportsThinking': supportsThinking,
        if (thinkingPattern != null)
          'thinkingPattern': thinkingPattern!.toJson(),
        if (metadata != null) 'metadata': metadata!.toJson(),
        'source': source.rawValue,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
        'syncPending': syncPending,
        if (lastUsed != null) 'lastUsed': lastUsed!.toIso8601String(),
        'usageCount': usageCount,
      };

  factory ModelInfo.fromJson(Map<String, dynamic> json) {
    return ModelInfo(
      id: json['id'] as String,
      name: json['name'] as String,
      category: ModelCategory.fromRawValue(json['category'] as String) ??
          ModelCategory.language,
      format: ModelFormat.fromRawValue(json['format'] as String),
      downloadURL: json['downloadURL'] != null
          ? Uri.parse(json['downloadURL'] as String)
          : null,
      localPath: json['localPath'] != null
          ? Uri.parse(json['localPath'] as String)
          : null,
      downloadSize: json['downloadSize'] as int?,
      memoryRequired: json['memoryRequired'] as int?,
      compatibleFrameworks: (json['compatibleFrameworks'] as List<dynamic>?)
              ?.map((f) => LLMFramework.fromRawValue(f as String))
              .whereType<LLMFramework>()
              .toList() ??
          [],
      preferredFramework: json['preferredFramework'] != null
          ? LLMFramework.fromRawValue(json['preferredFramework'] as String)
          : null,
      contextLength: json['contextLength'] as int?,
      supportsThinking: json['supportsThinking'] as bool? ?? false,
      thinkingPattern: json['thinkingPattern'] != null
          ? ThinkingTagPattern.fromJson(
              json['thinkingPattern'] as Map<String, dynamic>)
          : null,
      metadata: json['metadata'] != null
          ? ModelInfoMetadata.fromJson(json['metadata'] as Map<String, dynamic>)
          : null,
      source: ConfigurationSource.fromRawValue(
          json['source'] as String? ?? 'remote'),
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
      syncPending: json['syncPending'] as bool? ?? false,
      lastUsed: json['lastUsed'] != null
          ? DateTime.parse(json['lastUsed'] as String)
          : null,
      usageCount: json['usageCount'] as int? ?? 0,
    );
  }

  /// Copy with modifications
  ModelInfo copyWith({
    String? id,
    String? name,
    ModelCategory? category,
    ModelFormat? format,
    Uri? downloadURL,
    Uri? localPath,
    int? downloadSize,
    int? memoryRequired,
    List<LLMFramework>? compatibleFrameworks,
    LLMFramework? preferredFramework,
    int? contextLength,
    bool? supportsThinking,
    ThinkingTagPattern? thinkingPattern,
    ModelInfoMetadata? metadata,
    ConfigurationSource? source,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? syncPending,
    DateTime? lastUsed,
    int? usageCount,
    Map<String, String>? additionalProperties,
  }) {
    return ModelInfo(
      id: id ?? this.id,
      name: name ?? this.name,
      category: category ?? this.category,
      format: format ?? this.format,
      downloadURL: downloadURL ?? this.downloadURL,
      localPath: localPath ?? this.localPath,
      downloadSize: downloadSize ?? this.downloadSize,
      memoryRequired: memoryRequired ?? this.memoryRequired,
      compatibleFrameworks: compatibleFrameworks ?? this.compatibleFrameworks,
      preferredFramework: preferredFramework ?? this.preferredFramework,
      contextLength: contextLength ?? this.contextLength,
      supportsThinking: supportsThinking ?? this.supportsThinking,
      thinkingPattern: thinkingPattern ?? this.thinkingPattern,
      metadata: metadata ?? this.metadata,
      source: source ?? this.source,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      syncPending: syncPending ?? this.syncPending,
      lastUsed: lastUsed ?? this.lastUsed,
      usageCount: usageCount ?? this.usageCount,
      additionalProperties: additionalProperties ?? this.additionalProperties,
    );
  }

  /// Mark model as used
  void markAsUsed() {
    lastUsed = DateTime.now();
    usageCount++;
    updatedAt = DateTime.now();
    syncPending = true;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ModelInfo && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'ModelInfo(id: $id, name: $name, category: $category)';
}
